Add `Launch at Login` setting for macOS and Windows

whitecrownclown opened the pull request:
@sindresorhus 
Fixes #773 

Hello!
I've added the following : 

- Launch at login menu item
- Launch as hidden menu item (disabled if launch at login is false)
- Tested on macOS 10.14.3

Feedback is welcome, thanks!

sindresorhus commented:
You shouldn’t store the states in the config store. Just fetch the actual states from the `getLogin...` method when needed. That way the settings don’t get out of sync.

sindresorhus commented on source/menu.ts:
```diff
@@ -13,6 +13,23 @@ import config from './config';
 import {generateSubmenu as generateEmojiSubmenu} from './emoji';
 
+const appFolder = path.dirname(process.execPath);
+const updateExe = path.resolve(appFolder, '..', 'Update.exe');
```
?

whitecrownclown replied:
I only moved those outside of the function definitions since they're not updating at all

sindresorhus replied:
No, I mean. Can you explain why this is needed?

whitecrownclown replied:
Initially it was for : [To work with Electron's autoUpdater on Windows](https://electronjs.org/docs/api/app#appsetloginitemsettingssettings-macos-windows).

whitecrownclown replied:
I removed it for now, as i cannot really test it and also don't have access to windows.

sindresorhus commented on source/menu.ts:
```diff
@@ -190,6 +209,24 @@ Press Command/Ctrl+R in Caprine to see your changes.
 		},
+		{
+			type: 'checkbox',
+			label: 'Launch at login',
```
`label` should be first

sindresorhus commented:
> * Launch as hidden menu item (disabled if launch at login is false)

I don't see the point of having a setting for this. We should just launch hidden.

sindresorhus commented on source/menu.ts:
```diff
@@ -13,6 +13,23 @@ import config from './config';
+const appFolder = path.dirname(process.execPath);
+const updateExe = path.resolve(appFolder, '..', 'Update.exe');
+
+function setLoginSettings({openAtLogin, openAsHidden}: AppLoginSettings): void {
```
Just use the type for `setLoginItemSettings` instead of defining your own.

whitecrownclown replied:
Ok, i did look for it at first, didn't find it, i'll have a closer look.

whitecrownclown commented:
@sindresorhus I've updated the PR again.

* Added the ts definition from electron
* Removed `Launch as Hidden`; defaults to the same value as `Launch at Login`

sindresorhus commented on source/menu.ts:
```diff
@@ -13,6 +13,14 @@ import config from './config';
+
+function getLoginSettings(): LoginItemSettings {
+	return app.getLoginItemSettings();
+}
```
These wrappers are now moot.

whitecrownclown replied:
Yes, will update

sindresorhus commented on source/menu.ts:
```diff
@@ -190,6 +200,14 @@ Press Command/Ctrl+R in Caprine to see your changes.
+			type: 'checkbox',
+			checked: settings.openAtLogin,
+			click(item) {
+				setLoginSettings({...settings, openAtLogin: item.checked, openAsHidden: item.checked});
```
I don't think you need `...settings`. And `openAsHidden: item.checked` needs to be updated.

whitecrownclown replied:
I thought so as well, but i'm getting : `Type '{ openAtLogin: boolean; openAsHidden: boolean; }' is missing the following properties from type 'LoginItemSettings': wasOpenedAtLogin, wasOpenedAsHidden, restoreState`

whitecrownclown replied:
Disregard, it's no longer applicable since i've removed the wrappers.

whitecrownclown commented:
@sindresorhus Updated again, however i'm unsure what you mean by this:

>  And openAsHidden: item.checked needs to be updated.

Do you mean to set it to `true` always? It is being updated with the same value as `Launch at Login`. I've noticed if `Launch at Login` is `false` it will not set `Launch as Hidden` as `true`, but `false` as well.

sindresorhus commented:
> Do you mean to set it to `true` always?

Yes, setting it to `true` when `openAtLogin: false` should have no effect, so might as well make it simpler.

sindresorhus commented on source/menu.ts:
```diff
@@ -190,6 +190,14 @@ Press Command/Ctrl+R in Caprine to see your changes.
 			}
 		},
+		{
+			label: 'Launch at Login',
```
This feature only works on macOS and Windows, so you have to limit it to those system with the `visible` option.

https://electronjs.org/docs/api/app#appsetloginitemsettingssettings-macos-windows

whitecrownclown replied:
Yep, I missed that, thanks!

sindresorhus commented:
Thanks :)

blargle-ragequit commented:
Why is this not on Linux?

CvX commented:
@blargle-ragequit Because the Electron API doesn’t support it: https://electronjs.org/docs/api/app#appsetloginitemsettingssettings-macos-windows

Here’s an open issue at the Electron repo: https://github.com/electron/electron/issues/15198
